File: Sources/rust/src/game.rs
use crate::entity::Entity;
use crate::entity_builder::EntityBuilder;
use crate::geometry::Rect;

pub struct Game {
    entity_builder: EntityBuilder,
    pub bounds: Rect,
    pub entities: Vec<Entity>
}

impl Game {
    pub fn new(entity_builder: EntityBuilder, bounds: Rect) -> Self {
        Game {
            entity_builder,
            bounds,
            entities: Vec::new()
        }
    }
}

impl Game {
    pub fn update(&mut self, time_since_last_update: u128) {
        for entity in self.entities.iter_mut() {
            entity.update(time_since_last_update);
        }
    }

    pub fn add(&mut self, species: String) -> &Entity {
        let entity = self.entity_builder.entity(species);
        self.entities.push(entity);
        return self.entities.last().unwrap();
    }
}

File: Sources/cpp/src/game.cpp
#include "game.h"

#include <chrono>
#include <iostream>
#include <sstream>

#include "geometry.h"
#include "string_utils.h"

RenderedItem::RenderedItem(std::string spritePath, Rect frame) : spritePath(spritePath), frame(frame) {}

Game::Game(double fps) : fps(fps), entities(std::vector<Entity>({})) {}

void Game::update(std::chrono::milliseconds timeSinceLastUpdate) {
    std::lock_guard<std::mutex> lock(mtx);
    for (auto& entity : entities) {
        entity.update(timeSinceLastUpdate);
    }
}

Entity * Game::add(Entity entity) {
    std::lock_guard<std::mutex> lock(mtx);
    entities.push_back(entity);
    return &entities.back();
}

const uint32_t Game::numberOfEntities() {
    std::lock_guard<std::mutex> lock(mtx);
    return entities.size();
}

std::vector<RenderedItem> Game::render() {
    std::lock_guard<std::mutex> lock(mtx);
    std::vector<RenderedItem> renderedItems({});
    
    for (const auto& entity : entities) {
        auto item = RenderedItem(entity.currentSpriteFrame(), entity.frame);
        renderedItems.push_back(item);
    }
    return renderedItems;
}

std::string Game::description() {
    std::lock_guard<std::mutex> lock(mtx);
    std::stringstream ss; 

    ss << entities.size() << " entities:" << std::endl;

    for (const auto& entity : entities) {
        auto s = entity.description();
        trim(s);
        ss << "  - " << s << " @ " << &entity << std::endl;
    }
    return ss.str();
}